test/typescript/provider-upgrade-command/test.ts

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import { TestDriver, onPosix, onWindows } from "../../test-helper";
+
+describe("provider upgrade command", () => {
+  let driver: TestDriver;
+  beforeEach(async () => {
+    driver = new TestDriver(__dirname, {
+      CDKTF_DIST: "",
+      DISABLE_VERSION_CHECK: "true",
+      CI: "1",
+    }); // reset CDKTF_DIST set by run-against-dist script & disable version check as we have to use an older version of cdktf-cli
+    await driver.setupTypescriptProject({
+      init: { additionalOptions: "--cdktf-version 0.10.4" },
+    });
+  }, 500_000);
+
+  describe("pre-built", () => {
+    beforeEach(async () => {
+      await driver.exec("cdktf", [
+        "provider",
+        "add",
+        "random@=3.1.3", // this is not the latest version, but theres v0.2.55 of the pre-built provider resulting in exactly this package
+      ]);
+    });
+
+    it("can update withing the same cdktf version to a specific version", async () => {
+      expect(driver.npmPackageVersion("@cdktf/provider-random")).toEqual(
+        "^0.2.55"
+      );
+
+      await driver.exec("cdktf", ["provider", "upgrade", "random@=3.2.0"]);
+
+      expect(driver.npmPackageVersion("@cdktf/provider-random")).toEqual(
+        "^0.2.64"
+      );
+    });
+
+    it("can update withing the same cdktf version to the latest version", async () => {
+      expect(driver.npmPackageVersion("@cdktf/provider-random")).toEqual(
+        "^0.2.55"
+      );
+
+      await driver.exec("cdktf", ["provider", "upgrade", "random"]);
```

Comment:
Ah, nice! (If possible, I'd change the other tests to cover this case rather than the specific version one as it seems to be more common)

Reply:
I think it's sufficient to test this once, the algorithm is the same across the board. I'd rather have explicit versions in the test cases since it's clearer what is happening / expected to happen

Reply:
Yeah, makes sense for tests 👍